<a href="https://colab.research.google.com/github/wjlee62/Math-381-Wi26/blob/main/Homework_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Homework 1, due Friday, January 16**

You will submit two files for this assignment. Problems 1 and 2 should be completed by **hand**. Please submit a picture of your handwritten work (recommended) or submit a pdf of your typed soultions. You are **not** allowed to use a computer solver for Problems 1 and 2.

Problems 3-5 should be completed in this notebook, and you will submit the ipynb file as in Homework 0. There will be separate assignments on Gradescope for you to submit both files.

# Problems 1 and 2

1. **Winston, p. 63, Problem 5.**
2. **Winston, p. 113, Problem 1.** Describe all optimal solutions to the linear program model, and then state which of these solutions are realistic recommendations, assuming that fractions of barrels cannot be produced.

# Problem 3

Read **Winston, p. 115, Problem 13**. Then complete the following code to solve the problem. You must use the outline given. Note that the variable names cannot be chosen arbitrarily; you will need to look at the full code outline to see what the variables should be named.

To help the grader, you should **add comments** explaining your variables, constraints, etc. This will increase the chance that you get full credit.

This installs ortools for the rest of the notebook. Run this block.

In [7]:
%pip install ortools

The code for Problem 3 starts here.

In [8]:

from ortools.linear_solver import pywraplp

solver = pywraplp.Solver.CreateSolver("GLOP")

# Define the variables.
w1 = solver.NumVar(0,solver.infinity(), "wheat1")
w2 = solver.NumVar(0,solver.infinity(),"wheat2")
a1 = solver.NumVar(0,solver.infinity(),"alfalfa1")
a2 = solver.NumVar(0,solver.infinity(),"alfalfa2")

# Set the objective. Explain how you obtained it in the text box after the code block.
solver.Maximize(1*w1+1.1*a1+0.8*w2+0.9*a2)

# Add the constraints.
solver.Add(w1 >= 4*a1) #percentage of wheat in Feed 1
solver.Add(a2 >= 1.5*w2) #percentage of alfalfa in Feed 2
solver.Add(w1+w2 <= 1000) #wheat
solver.Add(a1+a2 <= 800) #alfalfa

# Solve and print the solution. DO NOT CHANGE ANYTHING PAST THIS POINT.
status = solver.Solve()

if status == pywraplp.Solver.OPTIMAL:
  print("Solution:")
  print(f"Optimal value = {solver.Objective().Value():0.2f}")
  print(f"{w1} = {w1.solution_value():0.2f}")
  print(f"{w2} = {w2.solution_value():0.2f}")
  print(f"{a1} = {a1.solution_value():0.2f}")
  print(f"{a2} = {a2.solution_value():0.2f}")
else:
  print("The problem does not have an optimal solution.")

Solution:
Optimal value = 1770.00
wheat1 = 1000.00
wheat2 = 0.00
alfalfa1 = 250.00
alfalfa2 = 550.00


**Explanation of objective function:**

The profit function can be written like this:
1.50*(w1+a1) + 1.30*(w2+a2)-0.5*(w1+w2)-0.4*(a1+a2)

This can be simplified to this:

(1.50-0.5)*w1 + (1.50-0.4)*a1 + (1.30-0.5)*w2 + (1.30-0.4)*a2 = 1.0*w1 + 1.1*a1 + 0.8*w2 + 0.9*a2

# Problem 4

Read **Winston, p. 115, Problem 14**. We will now modify the solver from Problem 3 to solve this new problem.

In [9]:
# Add four new variables to solver. You can choose the names (replace the XX's). The previous variables will still be in solver.
x1 = solver.NumVar(0, 300, "x1")
x2 = solver.NumVar(0, solver.infinity(), "x2")
y1 = solver.NumVar(0, 300, "y1")
y2 = solver.NumVar(0, solver.infinity(), "y2")

# Change the objective.
solver.Maximize(1.50*x1 + 1.25*x2 + 1.30*y1 + 1.00*y2 - 0.50*(w1 + w2) - 0.40*(a1 + a2))

# Add the new constraints. The previous constraints will still be in solver, and you should not change or remove them.
solver.Add(x1 + x2 == w1 + a1)
solver.Add(y1 + y2 == w2 + a2)

# Solve and print the solution. Change the XX's to match the variable names you chose.
status = solver.Solve()

if status == pywraplp.Solver.OPTIMAL:
  print("Solution:")
  print(f"Optimal value = {solver.Objective().Value():0.2f}")
  print(f"{w1} = {w1.solution_value():0.2f}")
  print(f"{w2} = {w2.solution_value():0.2f}")
  print(f"{a1} = {a1.solution_value():0.2f}")
  print(f"{a2} = {a2.solution_value():0.2f}")
  print(f"{x1} = {x1.solution_value():0.2f}")
  print(f"{x2} = {x2.solution_value():0.2f}")
  print(f"{y1} = {y1.solution_value():0.2f}")
  print(f"{y2} = {y2.solution_value():0.2f}")
else:
  print("The problem does not have an optimal solution.")

Solution:
Optimal value = 1457.50
wheat1 = 1000.00
wheat2 = 0.00
alfalfa1 = 250.00
alfalfa2 = 550.00
x1 = 300.00
x2 = 950.00
y1 = 300.00
y2 = 250.00


Run the block, and make sure it shows the output you want to submit.

# Problem 5

Read **Winston, p. 120, Problem 46**. Then complete the following code to solve the problem as a linear program.

In [10]:
solver = pywraplp.Solver.CreateSolver("GLOP")

# Complete the definitions of the variables. Use NumVar, not IntVar.
f12 = solver.NumVar(0, solver.infinity(), "f12") #full time 12-1
f1 = solver.NumVar(0, solver.infinity(), "f1") #full time 1-2
p9 = solver.NumVar(0, solver.infinity(), "p9") #part time 9-12
p10 = solver.NumVar(0, solver.infinity(), "p10") #part time 10-1
p11 = solver.NumVar(0, solver.infinity(), "p11") #part time 11-2
p12 = solver.NumVar(0, solver.infinity(), "p12") #part time 12-3
p1 = solver.NumVar(0, solver.infinity(), "p1") #part time 1-4
p2 = solver.NumVar(0, solver.infinity(), "p2") #part time 2-5

# Set the objective
solver.Minimize(64*(f12+f1) + 15*(p9 + p10 + p11 + p12 + p1 + p2))

# Add the constraints
solver.Add(f12 + f1 + p9 + p10 + p11 >= 4) #9-10 required tellers
solver.Add(f12 + f1 + p9 + p10 + p11 + p12 >= 3) #10-11 required tellers
solver.Add(f12 + f1 + p9 + p10 + p11 + p12 >= 4) #11-12 required tellers
solver.Add(f1 + p10 + p11 + p12 >= 6) #12-1 required tellers
solver.Add(f12 + p11 + p12 + p1 >= 5) #1-2 required tellers
solver.Add(f12 + f1 + p12 + p1 + p2 >= 6) #2-3 required tellers
solver.Add(f12 + f1 + p1 + p2 >= 8) #3-4 required tellers
solver.Add(f12 + f1 + p2 >= 8) #4-5 required tellers
solver.Add(p9 + p10 + p11 + p12 + p1 + p2 <=5) #limit in part time tellers

# Solve and print the solution. DO NOT CHANGE ANYTHING PAST THIS POINT.
status = solver.Solve()

if status == pywraplp.Solver.OPTIMAL:
  print("Solution:")
  print(f"Optimal value = {solver.Objective().Value():0.2f}")
  print(f"{f12} = {f12.solution_value():0.2f}")
  print(f"{f1} = {f1.solution_value():0.2f}")
  print(f"{p9} = {p9.solution_value():0.2f}")
  print(f"{p10} = {p10.solution_value():0.2f}")
  print(f"{p11} = {p11.solution_value():0.2f}")
  print(f"{p12} = {p12.solution_value():0.2f}")
  print(f"{p1} = {p1.solution_value():0.2f}")
  print(f"{p2} = {p2.solution_value():0.2f}")
else:
  print("The problem does not have an optimal solution.")

Solution:
Optimal value = 437.67
f12 = 2.33
f1 = 3.33
p9 = 0.00
p10 = 0.00
p11 = 0.00
p12 = 2.67
p1 = 0.00
p2 = 2.33


Based on the output of the program, give a realistic recommendation to the bank in the space provided. Give a brief explanation of how you came up with the recommendation. You do not need the prove that your recommendation is the best possible.

**Your recommendation:** Hire 6 full-time workers in total, where 2 of them take lunch from 12-1 and 4 of them take lunch from 1-2. Hire 5 part-time workers in total, where 3 of them work from 12-3 and 2 of them work from 2-5 (other part-time shifts have zero workers). This policy satisfies all the teller requirements and yields a labor cost that is very close to the optimal solution (437.67).

# Submit

Click "Run all" and make sure all output is shown. Download the .ipnyb file. Submit the file for Problems 1+2 and the .ipnyb file for Problems 3-5 to Gradescope.